contracts/Removal.sol

```diff
@@ -80,37 +110,87 @@ contract Removal is
     return removalId.unpackRemovalIdV0();
   }
 
+  // TODO do we want a batch version of this?
+  /**
+   * @notice Get the restriction schedule id (which is the removal's project id) for a given removal id.
+   */
+  function getProjectIdForRemoval(uint256 removalId)
+    public
+    view
+    returns (uint256)
+  {
+    return _removalIdToProjectId[removalId];
+  }
+
+  /**
+   * @notice Get the restriction schedule data for a given project id.
+   */
+  function getScheduleData(uint256 projectId)
+    public
+    view
+    returns (ScheduleData memory)
+  {
+    return _projectIdToScheduleData[projectId];
+  }
+
   /**
    * @dev mints multiple removals at once (for a single supplier).
    * If `list` is true in the decoded BatchMintRemovalsData, also lists those removals for sale in the market.
    * amounts: [100 * (10 ** 18), 10 * (10 ** 18), 50 * (10 ** 18)] <- 100 tonnes, 10 tonnes, 50 tonnes in standard erc20 units (wei)
    * token id 0 URI points to vintage information (e.g., 2018) nori.com/api/removal/0 -> { amount: 100, supplier: 1, vintage: 2018, ... }
    * token id 1 URI points to vintage information (e.g., 2019) nori.com/api/removal/1 -> { amount: 10, supplier: 1, vintage: 2019, ... }
    * token id 2 URI points to vintage information (e.g., 2020) nori.com/api/removal/2 -> { amount: 50, supplier: 1, vintage: 2020, ... }
+   *
    * @param to The supplier address
    * @param amounts Each removal's tonnes of CO2 formatted as wei
    * @param ids The token ids to use for this batch of removals. The id itself encodes the supplier's ethereum address, a parcel identifier,
-   * the vintage, country code, state code, methodology identifer, and methodology version.
-   * @param data Encodes the market contract address and a unique identifier for the parcel from whence these removals came.
+   * the vintage, country code, state code, methodology identifer, methodology version, and id format.
+   * @param data Encodes the project id and schedule start time for this batch of removals, the market contract
+   * address and a boolean that indicates whether to list these removals for sale now.
```

Comment:
FYI our massively minting test is now consuming 27_298_962 gas

Reply:
haha yeah I know we've got a problem there... it fails by blowing the block gas limit for me locally... not sure if CI has a different block gas limit configured?? because it's passing here.

Reply:
Here's my first attempt at a slightly gas optimized version which reduces that number by ~55,000 gas in our current worst case. Not sure what else we could optimize here, but the remaining bulk of the cost is coming from listing, super.mintBatch, and approvals (which is where 19 million of the 27 million gas is coming from)


```solidity
  function mintBatch(
    address to,
    uint256[] memory amounts,
    uint256[] memory ids,
    bytes memory data
  ) public override {
    uint256 idsLength = ids.length;
    if (!(amounts.length == idsLength)) {
      revert ArrayLengthMismatch({array1Name: "amounts", array2Name: "ids"});
    }
    // todo should we check that the removal id-encoded supplier address for each id is the same as `to` ?
    // todo should we validate the removal ids any further? what about other malformed fields?
    //    note that there is validation performed when creating ids in the first place, but room for errors/changes between id creation and then submission to this function
    // todo do we add any validation to enforce that all removals in batch belong to the same project id?
    BatchMintRemovalsData memory decodedData = abi.decode(
      data,
      (BatchMintRemovalsData)
    );
    uint256 projectId = decodedData.projectId;
    uint256 scheduleStartTime = decodedData.scheduleStartTime;
    bool list = decodedData.list;
    address marketAddress = decodedData.marketAddress;
    if (projectId == 0) {
      revert InvalidProjectId({projectId: projectId});
    }
    if (
      scheduleStartTime < UNIX_EPOCH_2010 || scheduleStartTime > UNIX_EPOCH_2050
    ) {
      revert InvalidScheduleStartTime({startTime: scheduleStartTime});
    }
    uint256 newTokenIdCounter = tokenIdCounter;
    for (uint256 i = 0; i < idsLength; i++) {
      uint256 id = ids[i];
      if (_tokenIdExists[id]) {
        revert TokenIdExists({tokenId: id});
      }
      _removalIdToProjectId[id] = projectId;
      _tokenIdExists[id] = true;
      indexToTokenId[newTokenIdCounter] = id;
      newTokenIdCounter += 1;
    }
    uint256 firstId = ids[0];
    tokenIdCounter = newTokenIdCounter;
    _projectIdToScheduleData[projectId] = ScheduleData(
      scheduleStartTime,
      firstId.supplierAddress(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
      firstId.methodology(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
      firstId.methodologyVersion() // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
    );
    super.mintBatch(to, ids, amounts, data);
    setApprovalForAll(to, _msgSender(), true); // todo look at vesting contract for potentially better approach
    if (list) {
      safeBatchTransferFrom(to, marketAddress, ids, amounts, data);
    }
  }
```